solved homework ArrayList

Yaromyros opened the pull request:
I tried to use all rules of good-looking code and all needed code-material. I override methods and created my own of increasation of array.

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double DEFAULT_GROWTH_OF_ARRAY = 1.5;
+    private Object[] elements = new Object[DEFAULT_CAPACITY];
```
```suggestion
    private T[] elements;
```
init instance fields (not constants) in constructor

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
     @Override
     public void add(T value) {
-
+        if (size == elements.length) {
+            increaseCapacity();
+        }
```
move `if` inside of this method

Vasul-Pankiv replied:
not fixed, move if statement into increaseCapacity method

Vasul-Pankiv replied:
Okay, i see that you already moved it into metho, so you no longer needed this if here, just remove it (same in other places where you call increase method)

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
+        if (size == elements.length) {
+            increaseCapacity();
+        }
+        elements[size] = value;
+        size++;
```
```suggestion
        elements[size++] = value;
```

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
+        size++;
+    }
 
+    private void increaseCapacity() {
+        int newCapacity = (int) (elements.length * DEFAULT_GROWTH_OF_ARRAY);
+        Object[] newElements = new Object[newCapacity];
+        System.arraycopy(elements, 0, newElements, 0, size);
+        elements = newElements;
     }
```
private methods should go after public ones

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds: " + index);
+        }
```
extract to a private method, include size in the message

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds: " + index);
+        }
+        T removedElement = (T) elements[index];
+        for (int i = index; i < size - 1; i++) {
+            elements[i] = elements[i + 1];
+        }
```
use system.arraycopy to move elements

okuzan commented on src/main/java/core/basesyntax/ArrayList.java:
```diff
@@ -1,48 +1,98 @@
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elements[i])) {
```
![image](https://github.com/mate-academy/jv-array-list/assets/49767049/985774be-4285-44f9-92b2-e43dadedce8a)

okuzan reviewed (changes requested):
Make sure to avoid checklist issues

Yaromyros commented:
All in all, I added constructor and solved some little mistakes. Moreover, I changed order of private and public methods, extracted one if statement to separated method. I checked on using system.arraycopy and prohibited classes.
